server/s3_get_object_action.cc

```diff
@@ -336,13 +336,72 @@ void S3GetObjectAction::read_object() {
   s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
 }
 
+void S3GetObjectAction::check_outbuffer_and_mempool_stats(bool& bcontinue) {
+  s3_log(S3_LOG_DEBUG, stripped_request_id, "%s Entry\n", __func__);
+  bcontinue = true;
+  size_t motr_write_payload_size =
+      S3Option::get_instance()->get_motr_write_payload_size(
+          object_metadata->get_layout_id());
+  // Determine the size of outstanding response buffer (not written to client
+  // sock)
+  size_t len_response_buffer = request->get_write_buffer_outstanding_length();
+  size_t len_mempool_free_mem = 0;
+  // Determine the size of outstanding mem pool buffer (free buffer)
+  struct pool_info poolinfo;
+  int rc = event_mempool_getinfo(&poolinfo);
+  if (rc != 0) {
+    s3_log(
+        S3_LOG_ERROR, request_id,
+        "Issue in reading memory pool stats during S3 Get API memory check\n");
+  } else {
+    len_mempool_free_mem =
+        poolinfo.free_bufs_in_pool * poolinfo.mempool_item_size;
+  }
+  s3_log(S3_LOG_INFO, request_id,
+         "Outstanding S3 Get response buffer size: (%zu)\n",
+         len_response_buffer);
+  s3_log(S3_LOG_INFO, request_id, "Free S3 mempool memory: (%zu)\n",
+         len_mempool_free_mem);
+  if ((len_response_buffer >=
+       (motr_write_payload_size *
+        S3Option::get_instance()->get_write_buffer_multiple())) ||
+      (!S3MemoryProfile().free_memory_in_pool_above_threshold_limits())) {
+    bcontinue = false;
+    s3_log(
+        S3_LOG_WARN, stripped_request_id,
+        "Limited memory: Throttling S3 GET object/part request is required\n");
+  }
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
 void S3GetObjectAction::read_object_data() {
   s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
   if (check_shutdown_and_rollback()) {
     s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
     return;
   }
-
+  // Before reading from Motr, ensure that outstanding response buffer
+  // is not above threshold1 or outstanding S3 mempool memory
+  // is not below threshold2
+  // threshold1 := motr_write_payload_size *
+  // S3Option::get_instance()->get_write_buffer_multiple()
+  // threshold2 :=
+  // S3MemoryProfile().free_memory_in_pool_above_threshold_limits()
+  bool bcontinue = true;
+  check_outbuffer_and_mempool_stats(bcontinue);
+  if (!bcontinue) {
+    // Throttle S3 Get API by adding delay using timer event
+    if (!request->set_start_response_delay_timer(
+             S3Option::get_instance()->get_s3_req_throttle_time(),
+             (void*)this)) {
+      s3_log(S3_LOG_INFO, request_id,
+             "S3 GET API response is throttled by: (%d) seconds\n", 3);
```

Comment:
have option value rather than hard coding

Reply:
S3 GET API will be ....

Reply:
throttle time in secs is option value through S3Option::get_instance()->get_s3_req_throttle_time()

Reply:
Removed '3' hardcoded value